Rockchip RK3128 Support

acmeplus opened the pull request:
Added support for several RK3128 based consoles:
* Powkiddy A12
* Powkiddy A13
* PS5000 (Also known as Q500 or Lintian LT-8203)
* PS7000 (Also known as Q900)

Other consoles like Powkiddy A9, Powkiddy X2, Supbor Q400, AIWO G-1000, etc. still require dedicated DTDs 
and may or may not work correctly. Additional information can be found [here](https://rk3128-cfw.github.io)

Known issues:
* Splash screen does not work, ``fbv`` throws an error related to mmap
* ``rngd`` takes forever and impacts CPU significantly
* GL driver does not perform correctly in retroarch, set SDL2 as your default driver in retroarch (note that bezels won't work with this driver)
* Initial audio volume may be too high for the Powkidddy A12, however after you lower and restart the system the new audio volume is saved
* Analogue controls are not working properly in PS5000/PS7000, see workarounds section below
* Digital controls are working fine on A12/A13 but not on PS5000/PS7000 since the keyboard mapping used with the digital controls conflicts with the default ``es_input.cfg`` mapping
* Remaping digital controls on PS5000/PS7000 works on ES, but controls are not properly mapped in retroarch (e.g. A becomes Z, etc.)
* Hotkey does not work correctly on PS5000/PS7000, see workarounds section below
* ``Moonlight-embedded`` has been disabled since its compilation currently fails due to a missing ``libcec`` include (libcec requires a patch to disable LINUX_API that does not exist in kernel 4.4.x). There's a workaround at buildroot level, but need to figure out how to add it at batocera level
* HDMI output is untested and probably not working at this point
* Audio jack audio/switch is untested and probably not working at this point 

Workarounds:
* PS5000/PS7000 analogue controls, add the following lines at the end of ``batocera.conf``
``` console
global.retroarch.input_player1_l_x_minus_axis = "-0"
global.retroarch.input_player1_l_x_plus_axis = "+0"
global.retroarch.input_player1_l_y_minus_axis = "-1"
global.retroarch.input_player1_l_y_plus_axis = "+1"
global.retroarch.input_player1_r_x_minus_axis = "+2"
global.retroarch.input_player1_r_x_plus_axis = "-2"
global.retroarch.input_player1_r_y_minus_axis = "+3"
global.retroarch.input_player1_r_y_plus_axis = "-3"
```

* PS5000/PS7000 hotkey not working, add the following line to ``/userdata/systems/batocera.conf``:
``global.retroarch.input_menu_toggle_gamepad_combo = "4"``

* PS5000/PS7000 analogue joystick not working on ES, add the following lines to ``configs/emulationstation/es_input.cfg``:
```xml
<inputConfig type="joystick" deviceName="WX8 joysticks" deviceGUID="18000000575838206a6f797374696300">
		<input name="joystick1left" type="axis" id="0" value="-1" code="0" />
		<input name="joystick1up" type="axis" id="1" value="-1" code="1" />
		<input name="joystick2left" type="axis" id="2" value="-1" code="2" />
		<input name="joystick2up" type="axis" id="3" value="-1" code="5" />
</inputConfig>
```

* Getting Bezels back in case you are curious (note performance is significantly impacted). Add the following line at the end of ``batocera.conf``: ``global.retroarch.video_driver="gl"``

nadenislamarre commented:
very nice.
1) detail : the yml desn't need to be so long. for example, rpcs3 can be removed. it is the delta from the default.
2) 
if(type == DRM_MODE_CONNECTOR_DPI) 	     return "DPI";
  //if(type == DRM_MODE_CONNECTOR_DPI) 	     return "DPI";

why ?
if really needed, just do a patch in board/batocera/rk3128/patches/batocera-drminfo to apply it only for these boards

3)
-default "rockchip/rg552"                                        if BR2_PACKAGE_BATOCERA_TARGET_RG552
why did you remove the rg552 board ?

4)
you commited a bad file : /ja_JP/batocera-es-system.po~

5)
[package/batocera/gpu/mali-mp400-gbm/000-add-EGL_NO_X11_directive.patch.disabled](https://github.com/batocera-linux/batocera.linux/pull/5839/files#diff-6158968e71c6c0078b566f82f9884ac37704a64ed03440c3b747d2bddb9f51bb)
why did you add this file ?

for the rest, it is ok to me.
so, only minor modifications required.
i'm not fan of all the .sh you put on fsoverlay, but possibly a good reason and for the moment, it is ok.
after that step, we can take point by point your issue so that i suggest some way to do.

acmeplus commented:
PR updated, some comments.

> very nice.
> 
> 1. detail : the yml desn't need to be so long. for example, rpcs3 can be removed. it is the delta from the default.

Done, just left the defaults, assume that should be enough.

> if(type == DRM_MODE_CONNECTOR_DPI) return "DPI"; //if(type == DRM_MODE_CONNECTOR_DPI) return "DPI";
> 
> why ? if really needed, just do a patch in board/batocera/rk3128/patches/batocera-drminfo to apply it only for these boards

I've reverted the change although RK3128 build is currently broken. The issue is I don't know how to create a patch that patches something at the package level, because ``batocera-drminfo.c`` lives in ``package/batocera/core/batocera-drminfo/``. 

> -default "rockchip/rg552" if BR2_PACKAGE_BATOCERA_TARGET_RG552 why did you remove the rg552 board ?

Done, added it back, it was a merge mistake
 
> you commited a bad file : /ja_JP/batocera-es-system.po~

Done, removed.

> [package/batocera/gpu/mali-mp400-gbm/000-add-EGL_NO_X11_directive.patch.disabled](https://github.com/batocera-linux/batocera.linux/pull/5839/files#diff-6158968e71c6c0078b566f82f9884ac37704a64ed03440c3b747d2bddb9f51bb) why did you add this file ?

Removed, no longer needed. 
 
> for the rest, it is ok to me. so, only minor modifications required. i'm not fan of all the .sh you put on fsoverlay, but possibly a good reason and for the moment, it is ok. after that step, we can take point by point your issue so that i suggest some way to do.

Will most likely remove all those .sh files from fsoverlay once I have more tests results. Those were really for hdmi and audio switching in a buildroot only build, but I suspect they are mostly irrelevant since batocera manages those differently.

nadenislamarre commented:
seems you have an empty @ file too, but anyway, let's me recheck, merge and remove this file.
then, i run a compilation, we discuss each point you have to fix one by one

nadenislamarre commented:
# Misc
# BR2_PACKAGE_XOW is not set 
BR2_PACKAGE_LIBCEC=y
BR2_PACKAGE_INPUT_EVENT_DAEMON=y
BR2_PACKAGE_LIBSODIUM=y

is probably to not put in the _defconfig.
but in the package/batocera/core/Config.in

nadenislamarre commented:
merged. i suggest you create an issue, listing points remaining. and we take them one by one.
(something like : https://github.com/batocera-linux/batocera.linux/issues/5518)

nadenislamarre commented:
drminfo fixed and commited. beta built. (no way to test on my side)

acmeplus commented:
> merged. i suggest you create an issue, listing points remaining. and we take them one by one.
> (something like : #5518)

Thanks! will open an issue in a bit and list all the points there.

acmeplus commented:
Added new #5845 to collect all the current  issues detected
